docs/pre-submission-hooks.md

```diff
@@ -0,0 +1,259 @@
+
+**Parameters:**
+
+  - `pipeline.var_templates.sample_yaml_path` (optional): absolute path to file where YAML is to be stored.
+
+**Usage:**
+
+```yaml
+pipeline_type: sample
+var_templates:
+  main: "{looper.piface_dir}/pipelines/pipeline1.py"
+  sample_yaml_cwl_path: "{looper.output_dir}/custom_sample_yamls/custom_{sample.name}.yaml"
+pre_submit:
+  python_functions:
+    - looper.write_sample_yaml_cwl
+command_template: >
+  {pipeline.var_templates.main} ...
+```
+
+
+### Included plugin: `looper.write_sample_yaml_prj`
+
+Saves the sample to YAML file with project reference.  This plugin can be parametrized with a custom YAML directory (see "parameters" below). If the parameter is not provided, the file will be saved in `{looper.output_dir}/submission/{sample.sample_name}_sample_prj.yaml`.
+
+**Parameters:**
+  
+  - `pipeline.var_templates.sample_yaml_prj_path` (optional): absolute path to file where YAML is to be stored.
+
+**Usage:**
+
+```yaml
+pipeline_type: sample
+var_templates:
+  main: "{looper.piface_dir}/pipelines/pipeline1.py"
+  sample_yaml_prj_path: "{looper.output_dir}/custom_sample_yamls"
+pre_submit:
+  python_functions:
+    - looper.write_sample_yaml_prj
+command_template: >
+  {pipeline.var_templates.main} ...
+```
+
+### Included plugin: `looper.write_submission_yaml`
+
+Saves all five namespaces of pre-submission to YAML file.  This plugin can be parametrized with a custom YAML directory (see "parameters" below). If the parameter is not provided, the file will be saved in `{looper.output_dir}/submission/{sample.sample_name}_submission.yaml`.
+
+**Parameters:**
+  
+  - `pipeline.var_templates.submission_yaml_path` (optional): a complete and absolute path to the *directory* where submission YAML representation is to be stored.
+
+**Example usage:**
+
+```yaml
+pipeline_type: sample
+var_templates:
+  main: "{looper.piface_dir}/pipelines/pipeline1.py"
+  submission_yaml_path: "{looper.output_dir}/custom_path"
+pre_submit:
+  python_functions:
+    - looper.write_submission_yaml
+command_template: >
+  {pipeline.var_templates.main} ...
+```
+
+## Writing your own pre-submission hooks
+
+Pre-submission tasks can be written as a Python function or a shell commands. We will explain each type below:
+
+### Python functions
+
+Python plugin functions have access *all of the metadata variables looper has access to to construct the primary command template*. The Python function must obey the following rules:
+
+1. The Python function *must* take as input a `namespaces` object, which is a Python [`dict`](https://docs.python.org/3/tutorial/datastructures.html#dictionaries) of [looper variable namespaces](variable-namespaces.md). 
+
+2. The function *should* return any updated namespace variables; or can potentially return `None` if no changes are made, which may the case if the function is only used for its side effect.
```

Comment:
@nsheff just realized this is not true, returning `None` would lead to an error, both in python_function and command_template hooks. Both need to return a `{}` if no changes are intended. Should I adjust the docs or implement `None` handling?